From 05067d9a1b07d12fe6aaafd3f1c1ff645f01d9f9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20=C5=BBygowski?= <michal.zygowski@3mdeb.com>
Date: Fri, 2 Oct 2026 12:20:21 +0200
Subject: [PATCH 4/4] Platform/Sgi: enable the EL3 fTPM and TCG2 measured boot
 on RD-V3

Use EDK2_ENABLE_EL3_FTPM to set PcdEl3FtpmEnable and add the TCG2 stack
(El3FtpmConfigPei, Tcg2Pei, Tcg2Dxe, Tcg2ConfigDxe), measured boot of
PE/COFF images and the GPT, the TPM2 ACPI tables, and map the CRB pages as
device memory. Set the PcdTpm2HashMask to extend SHA-256 and SHA-384, the
banks the Microsoft reference TPM code allocates by default.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
---
 Platform/ARM/SgiPkg/El3Ftpm.dsc.inc           | 81 +++++++++++++++++++
 .../Library/PlatformLib/PlatformLib.inf       |  1 +
 .../Library/PlatformLib/PlatformLibMem.c      | 11 +++
 Platform/ARM/SgiPkg/RdV3/RdV3.dsc             | 12 +++
 Platform/ARM/SgiPkg/RdV3/RdV3.fdf.inc         |  8 ++
 Platform/ARM/SgiPkg/SgiPlatform.dec           |  2 +
 Platform/ARM/SgiPkg/SgiPlatform.dsc.inc       | 11 ++-
 Platform/ARM/SgiPkg/SgiPlatform.fdf           |  4 +
 8 files changed, 127 insertions(+), 3 deletions(-)
 create mode 100644 Platform/ARM/SgiPkg/El3Ftpm.dsc.inc

diff --git a/Platform/ARM/SgiPkg/El3Ftpm.dsc.inc b/Platform/ARM/SgiPkg/El3Ftpm.dsc.inc
new file mode 100644
index 00000000..27020f25
--- /dev/null
+++ b/Platform/ARM/SgiPkg/El3Ftpm.dsc.inc
@@ -0,0 +1,81 @@
+## @file
+#  TCG2 (TPM 2.0) support for the TF-A EL3 firmware TPM (proof of concept).
+#
+#  The TPM is reached through the CRB interface with the Arm SMC start
+#  method, see Library/Tpm2DeviceLibCrbSmc. Requires TF-A built with
+#  NRD_EL3_FTPM=1.
+#
+#  Copyright (c) 2025, 3mdeb. All rights reserved.
+#
+#  SPDX-License-Identifier: BSD-2-Clause-Patent
+##
+
+[LibraryClasses.common]
+  Tpm2DeviceLib|Platform/ARM/SgiPkg/Library/Tpm2DeviceLibCrbSmc/Tpm2DeviceLibCrbSmc.inf
+  Tpm2CommandLib|SecurityPkg/Library/Tpm2CommandLib/Tpm2CommandLib.inf
+  Tpm12DeviceLib|SecurityPkg/Library/Tpm12DeviceLibDTpm/Tpm12DeviceLibDTpm.inf
+  Tpm12CommandLib|SecurityPkg/Library/Tpm12CommandLib/Tpm12CommandLib.inf
+  Tcg2PpVendorLib|SecurityPkg/Library/Tcg2PpVendorLibNull/Tcg2PpVendorLibNull.inf
+  TpmPlatformHierarchyLib|SecurityPkg/Library/PeiDxeTpmPlatformHierarchyLibNull/PeiDxeTpmPlatformHierarchyLib.inf
+
+[LibraryClasses.common.PEIM]
+  BaseCryptLib|CryptoPkg/Library/BaseCryptLib/PeiCryptLib.inf
+  Tcg2PhysicalPresenceLib|SecurityPkg/Library/PeiTcg2PhysicalPresenceLib/PeiTcg2PhysicalPresenceLib.inf
+
+[LibraryClasses.common.DXE_DRIVER, LibraryClasses.common.UEFI_DRIVER, LibraryClasses.common.UEFI_APPLICATION]
+  Tcg2PhysicalPresenceLib|SecurityPkg/Library/DxeTcg2PhysicalPresenceLib/DxeTcg2PhysicalPresenceLib.inf
+
+[LibraryClasses.common.DXE_DRIVER, LibraryClasses.common.DXE_RUNTIME_DRIVER, LibraryClasses.common.UEFI_DRIVER, LibraryClasses.common.UEFI_APPLICATION]
+  # Let other drivers (e.g. variable services) log measurements through TCG2
+  TpmMeasurementLib|SecurityPkg/Library/DxeTpmMeasurementLib/DxeTpmMeasurementLib.inf
+
+[PcdsFeatureFlag.common]
+  gArmSgiTokenSpaceGuid.PcdEl3FtpmEnable|TRUE
+
+[PcdsFixedAtBuild.common]
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpmBaseAddress|0xF7400000
+  # The CRB/SMC interface cannot be probed, use the configured device
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpmAutoDetection|FALSE
+
+[PcdsDynamicDefault.common]
+  # Send TPM2_Startup in PEI (also written by Tcg2ConfigDxe)
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpmInitializationPolicy|1
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpm2InitializationPolicy|1
+  # gEfiTpmDeviceInstanceTpm20DtpmGuid
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpmInstanceGuid|{0x5a, 0xf2, 0x6b, 0x28, 0xc3, 0xc2, 0x8c, 0x40, 0xb3, 0xb4, 0x25, 0xe6, 0x75, 0x8b, 0x73, 0x17}
+  #
+  # Must match the banks the fTPM allocates: SHA256 and SHA384 (the
+  # ms-tpm-20-ref default since v1.83; SHA1 is supported but not allocated).
+  # A bank allocated by the TPM but missing here makes Tcg2Pei reallocate the
+  # PCR banks and reset the platform, which loops forever because the fTPM
+  # state is volatile. A bank set here but not allocated is only dropped from
+  # every event ("Event log has HashAlg unsupported by PCR bank").
+  #
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpm2HashMask|0x00000006
+  gEfiSecurityPkgTokenSpaceGuid.PcdTcg2HashAlgorithmBitmap|0
+  gEfiSecurityPkgTokenSpaceGuid.PcdTcgPhysicalPresenceInterfaceVer|"1.3"
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpm2AcpiTableRev|4
+
+[Components.common]
+  #
+  # PEI
+  #
+  # Replaces Tcg2ConfigPei, which would select a TPM 1.2 device
+  Platform/ARM/SgiPkg/Drivers/El3FtpmConfigPei/El3FtpmConfigPei.inf
+  SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.inf {
+    <LibraryClasses>
+      HashLib|SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterPei.inf
+      NULL|SecurityPkg/Library/HashInstanceLibSha256/HashInstanceLibSha256.inf
+      NULL|SecurityPkg/Library/HashInstanceLibSha384/HashInstanceLibSha384.inf
+  }
+
+  #
+  # DXE
+  #
+  SecurityPkg/Tcg/Tcg2Dxe/Tcg2Dxe.inf {
+    <LibraryClasses>
+      HashLib|SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.inf
+      NULL|SecurityPkg/Library/HashInstanceLibSha256/HashInstanceLibSha256.inf
+      NULL|SecurityPkg/Library/HashInstanceLibSha384/HashInstanceLibSha384.inf
+  }
+  SecurityPkg/Tcg/Tcg2Config/Tcg2ConfigDxe.inf
diff --git a/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLib.inf b/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLib.inf
index 82355de8..62ee05db 100644
--- a/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLib.inf
+++ b/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLib.inf
@@ -99,6 +99,7 @@
 
 [FeaturePcd]
   gArmSgiTokenSpaceGuid.PcdEinjSupported
+  gArmSgiTokenSpaceGuid.PcdEl3FtpmEnable
 
 [Guids]
   gArmSgiPlatformIdDescriptorGuid
diff --git a/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLibMem.c b/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLibMem.c
index 35210285..66ae4dc4 100644
--- a/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLibMem.c
+++ b/Platform/ARM/SgiPkg/Library/PlatformLib/PlatformLibMem.c
@@ -16,9 +16,12 @@
 #include <IoVirtSoCExp.h>
 #include <SgiPlatform.h>
 
+#include <RdV3El3Ftpm.h>
+
 // Total number of descriptors, including the final "end-of-table" descriptor.
 #define MAX_VIRTUAL_MEMORY_MAP_DESCRIPTORS                                                  \
           ((14 + (FixedPcdGet32 (PcdChipCount) * 2)) +                                      \
+          (FeaturePcdGet (PcdEl3FtpmEnable)) +                                              \
           (FeaturePcdGet (PcdEinjSupported)) +                                              \
            ((FeaturePcdGet (PcdPcieEnable) == TRUE) ? (FixedPcdGet32 (PcdChipCount)) : 0) +  \
            (FixedPcdGet32 (PcdIoVirtSocExpBlkUartEnable) *                                  \
@@ -332,6 +335,14 @@ ArmPlatformGetVirtualMemoryMap (
     VirtualMemoryTable[Index].Attributes      = ARM_MEMORY_REGION_ATTRIBUTE_UNCACHED_UNBUFFERED;
   }
 
+  if (FeaturePcdGet (PcdEl3FtpmEnable)) {
+    // TF-A EL3 firmware TPM: CRB control area and command/response buffer
+    VirtualMemoryTable[++Index].PhysicalBase  = RDV3_EL3_FTPM_CRB_BASE;
+    VirtualMemoryTable[Index].VirtualBase     = RDV3_EL3_FTPM_CRB_BASE;
+    VirtualMemoryTable[Index].Length          = RDV3_EL3_FTPM_CRB_SIZE;
+    VirtualMemoryTable[Index].Attributes      = ARM_MEMORY_REGION_ATTRIBUTE_DEVICE;
+  }
+
   // End of Table
   VirtualMemoryTable[++Index].PhysicalBase  = 0;
   VirtualMemoryTable[Index].VirtualBase     = 0;
diff --git a/Platform/ARM/SgiPkg/RdV3/RdV3.dsc b/Platform/ARM/SgiPkg/RdV3/RdV3.dsc
index 4b6a3e5f..0f1739f0 100644
--- a/Platform/ARM/SgiPkg/RdV3/RdV3.dsc
+++ b/Platform/ARM/SgiPkg/RdV3/RdV3.dsc
@@ -61,4 +61,16 @@ DEFINE PCIE_ENABLE             = TRUE
 ################################################################################
 
 [Components.common]
+!if $(EDK2_ENABLE_EL3_FTPM) == TRUE
+  Platform/ARM/SgiPkg/AcpiTables/RdV3AcpiTablesEl3Ftpm.inf
+!else
   Platform/ARM/SgiPkg/AcpiTables/RdV3AcpiTables.inf
+!endif
+
+#
+# TF-A EL3 firmware TPM (PoC, TF-A NRD_EL3_FTPM=1): ACPI description and the
+# TCG2 driver stack
+#
+!if $(EDK2_ENABLE_EL3_FTPM) == TRUE
+!include Platform/ARM/SgiPkg/El3Ftpm.dsc.inc
+!endif
diff --git a/Platform/ARM/SgiPkg/RdV3/RdV3.fdf.inc b/Platform/ARM/SgiPkg/RdV3/RdV3.fdf.inc
index 89140426..8f9cf6bc 100644
--- a/Platform/ARM/SgiPkg/RdV3/RdV3.fdf.inc
+++ b/Platform/ARM/SgiPkg/RdV3/RdV3.fdf.inc
@@ -7,4 +7,12 @@
 # Per-platform additional content of the DXE phase firmware volume
 
   # ACPI support
+!if $(EDK2_ENABLE_EL3_FTPM) == TRUE
+  INF RuleOverride=ACPITABLE Platform/ARM/SgiPkg/AcpiTables/RdV3AcpiTablesEl3Ftpm.inf
+
+  # TCG2 (TPM 2.0) support for the EL3 firmware TPM
+  INF SecurityPkg/Tcg/Tcg2Dxe/Tcg2Dxe.inf
+  INF SecurityPkg/Tcg/Tcg2Config/Tcg2ConfigDxe.inf
+!else
   INF RuleOverride=ACPITABLE Platform/ARM/SgiPkg/AcpiTables/RdV3AcpiTables.inf
+!endif
diff --git a/Platform/ARM/SgiPkg/SgiPlatform.dec b/Platform/ARM/SgiPkg/SgiPlatform.dec
index 6a79686f..8a9c13c0 100644
--- a/Platform/ARM/SgiPkg/SgiPlatform.dec
+++ b/Platform/ARM/SgiPkg/SgiPlatform.dec
@@ -34,6 +34,8 @@
   gArmSgiTokenSpaceGuid.PcdVirtioNetSupported|FALSE|BOOLEAN|0x00000010
   gArmSgiTokenSpaceGuid.PcdPcieEnable|FALSE|BOOLEAN|0x0000002E
   gArmSgiTokenSpaceGuid.PcdEinjSupported|FALSE|BOOLEAN|0x00001001
+  # TF-A EL3 firmware TPM (PoC) with a CRB interface and SMC start method
+  gArmSgiTokenSpaceGuid.PcdEl3FtpmEnable|FALSE|BOOLEAN|0x00001200
 
 [PcdsFixedAtBuild]
 
diff --git a/Platform/ARM/SgiPkg/SgiPlatform.dsc.inc b/Platform/ARM/SgiPkg/SgiPlatform.dsc.inc
index bfe87442..3eb3146e 100644
--- a/Platform/ARM/SgiPkg/SgiPlatform.dsc.inc
+++ b/Platform/ARM/SgiPkg/SgiPlatform.dsc.inc
@@ -36,6 +36,7 @@
 
   # To enable FF-A as communication protocol between SP and Non secure world
   DEFINE EDK2_ENABLE_FFA                   = FALSE
+  DEFINE EDK2_ENABLE_EL3_FTPM              = FALSE
 
   # To enable drivers required for SPMC running at EL3
   DEFINE EDK2_SPMC_AT_EL2                  = FALSE
@@ -308,14 +309,18 @@
   MdeModulePkg/Universal/CapsuleRuntimeDxe/CapsuleRuntimeDxe.inf
   MdeModulePkg/Universal/MonotonicCounterRuntimeDxe/MonotonicCounterRuntimeDxe.inf
   MdeModulePkg/Universal/ResetSystemRuntimeDxe/ResetSystemRuntimeDxe.inf
-!if $(SECURE_BOOT_ENABLE) == TRUE
   MdeModulePkg/Universal/SecurityStubDxe/SecurityStubDxe.inf {
     <LibraryClasses>
+!if $(SECURE_BOOT_ENABLE) == TRUE
       NULL|SecurityPkg/Library/DxeImageVerificationLib/DxeImageVerificationLib.inf
+!endif
+!if $(EDK2_ENABLE_EL3_FTPM) == TRUE
+      # Measured boot: PE/COFF images into PCR 2/4, GPT into PCR 5
+      NULL|SecurityPkg/Library/DxeTpm2MeasureBootLib/DxeTpm2MeasureBootLib.inf
+!endif
   }
+!if $(SECURE_BOOT_ENABLE) == TRUE
   SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigDxe.inf
-!else
-  MdeModulePkg/Universal/SecurityStubDxe/SecurityStubDxe.inf
 !endif
   OvmfPkg/VirtioBlkDxe/VirtioBlk.inf
 
diff --git a/Platform/ARM/SgiPkg/SgiPlatform.fdf b/Platform/ARM/SgiPkg/SgiPlatform.fdf
index 47551d39..59db9136 100644
--- a/Platform/ARM/SgiPkg/SgiPlatform.fdf
+++ b/Platform/ARM/SgiPkg/SgiPlatform.fdf
@@ -261,6 +261,10 @@ READ_LOCK_STATUS   = TRUE
   INF MdeModulePkg/Universal/PCD/Pei/Pcd.inf
   INF MdeModulePkg/Universal/Variable/Pei/VariablePei.inf
   INF Platform/ARM/SgiPkg/Library/SgiPlatformPei/SgiPlatformPei.inf
+!if $(EDK2_ENABLE_EL3_FTPM) == TRUE
+  INF Platform/ARM/SgiPkg/Drivers/El3FtpmConfigPei/El3FtpmConfigPei.inf
+  INF SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.inf
+!endif
 
   FILE FV_IMAGE = 9E21FD93-9C72-4c15-8C4B-E77F1DB2D792 {
     SECTION GUIDED EE4E5898-3914-4259-9D6E-DC7BD79403CF PROCESSING_REQUIRED = TRUE {
